daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala

```diff
@@ -138,6 +147,22 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
       case Right(x) => x
     }
 
+  private[this] def decodeInternedTypes(
+      env: DecoderEnv,
+      lfTypes: Seq[PLF.Type],
+  ): IndexedSeq[Type] = {
+    if (lfTypes.nonEmpty)
+      assertSince(LV.Features.internedTypes, "interned types table")
+    val internedTypes = Array.ofDim[Type](lfTypes.length)
+    var i = 0
+    while (i < internedTypes.length) {
+      val typ = env.copy(internedTypes = internedTypes.view.slice(0, i)).decodeType(lfTypes(i))
+      internedTypes(i) = typ
+      i += 1
+    }
+    internedTypes
+  }
```

Comment:
@remyhaemmerle-da @S11001001 This implementation is probably very bad Scala but I honestly don't know how to do better. The Scala collections library is just confusing me. I'd be interested to improve it under your guidance.

Reply:
I would just do a fold.

```scala
    lfTypes.foldLeft(Vector.empty[Type]){ (internedTypes, lfType) =>
      val typ = env.copy(internedTypes = internedTypes).decodeType(lfType)
      internedTypes :+ typ
    }
```

Reply:
Thank you. I assume `Vector` is a mutable data structure with amortized `O(1)` pushing onto the end. Is that correct? If so, could I also take advantage of the fact that I know the final size by reserving that as the initial capacity?

Reply:
No, `Vector` is immutable with cheap snoc. https://docs.scala-lang.org/overviews/collections/concrete-immutable-collection-classes.html#vectors

Reply:
Would you still recommend using `Vector` even if I told you that there might be more than 300,000 different entries in the type interning table?

Reply:
Yes.

Reply:
I don't really like the idea of using immutable data structures for this at that size and since the mutation is _very_ local, I've decided to use the `ArrayBuffer` solution proposed by @remyhaemmerle-da.